File: app/src/app/services/podcast.clj
(ns app.services.podcast
  (:require [clj-http.client :as httpclient]
            [feedparser-clj.core :as feed-parser]))

(defn- get-title
  [f]
  (:title f))

(defn- get-entries
  [f]
  (:entries f))

(defn- get-description
  [f]
  (:description f))

(defn- get-logo
  [f]
  (:image f))

(defn- build-item
  [e]
  {
    :title          (:title e)
    :description    (:value (:description e))
    :url            (:url (first (:enclosures e)))
    :publishedDate  (:published-date e)
    :duration       (:length (first (:enclosures e)))
    })

(defn- get-items
  [f]
  (map build-item (get-entries f)))

(defn- get-copyright
  [f]
  (:copyright f))

(defn parse-feed
  [feed-url]
  (let [feed (feed-parser/parse-feed feed-url)]
    {
      :title        (get-title feed)
      :logo         (get-logo feed)
      :description  (get-description feed)
      :copyright    (get-copyright feed)
      :items        (get-items feed)
    }))

File: public/src/app/shared/podcast/podcast.ts
export class Podcast {
  public active: boolean;
  public description: string;
  public feed: PodcastFeed;
  public genre: string;
  public id: number;
  public name: string;
  public releaseDate: string;

  constructor(podcast?: Podcast) {
    if (podcast) {
      this.active = podcast.active;
      this.description = podcast.description;
      this.feed = new PodcastFeed(podcast.feed);
      this.genre = podcast.genre;
      this.id = podcast.id;
      this.name = podcast.name;
      this.releaseDate = podcast.releaseDate;
    }
  }
}

export class PodcastFeed {
  public copyright: string;
  public description: string;
  public items: PodcastEpisode[];
  public logo: string;
  public title: string;
  constructor(feed?: PodcastFeed) {
    if (feed) {
      this.copyright = feed.copyright;
      this.description = feed.description;
      this.items = [];
      feed.items.map((c) => {
        this.items.push(new PodcastEpisode(c));
      });
      this.logo = feed.logo;
      this.title = feed.title;
    }
  }
}

export class PodcastEpisode {
  public description: string;
  public duration: number;
  public publishedDate: string;
  public title: string;
  public url: string;
  public playing?: boolean = false;
  public loaded?: boolean = false;

  constructor(episode?: PodcastEpisode) {
    if (episode) {
      this.description = episode.description;
      this.duration = episode.duration;
      this.publishedDate = episode.publishedDate;
      this.title = episode.title;
      this.url = episode.url;
      this.playing = false;
      this.loaded = false;
    }
  }
}
